### DISCUSSION TITLE
Assigning Value of a Variable for Data to a Global Variable

### DISCUSSION THREAD

**Discussion by grkal** [groops-devs/groops#192] [Q&A] [answered]
Dear GROOPS Team,

The topic of this message concerns the possibility of assigning the value of a variable for data (as described in subchapter “1.2.3 Variables for data” of the GROOPS documentation) to a global variable.

Is it possible to somehow assign the value of a variable of data, e.g. the mean value of the X column of an input instrument file, to a global variable instead of saving this value only in an output file? For certain processing purposes of instrument files it would be helpful if this can be achieved.

Best regards,
Grigorios Kalimeris
Chair of Astronomical and Physical Geodesy
Technical University of Munich

**Comment by tmayerguerr** [comment, marked as answer]
Dear Grigorios,

A possible trick is to create a file with the values and when setting up a [loop](https://groops-devs.github.io/groops/html/loopType.html)->matrix (e.g. with LoopProgram) read the data and assign it to variables.

regards
Torsten

**Comment by grkal** [comment]
Dear Torsten,

Thank you very much for providing this very interesting solution. I also thought about using a [loop](https://groops-devs.github.io/groops/html/loopType.html) after I remembered that it also accepts a matrix as input.

On a related note, is it somehow possible to extract the number of columns of a matrix? As I have seen, there is not a [variable for data](https://groops-devs.github.io/groops/html/general.parser.html#dataVariables) which corresponds to the number of columns.

Best regards and I really appreciate your help,
Greg

**Reply by grkal** [reply]
Dear Torsten,

As a follow-up to my question regarding which way a matrix's number of columns could be extracted from, I found two possible solutions.

The first one is a combination of the "MatrixCalculate" and "FunctionsCalculate" programs. Specifically, the "MatrixCalculate" program with the "transpose" option and as "inputfile" = (the file containing the matrix from which the columns will be extracted) can be used initially so as to derive the transposed matrix. Subsequently, by using the transposed matrix as input for the "FunctionsCalculate" program with the "statistics" config element set to "< enabled >", "outputfile" = (the file where the columns will be stored - can be the same file name with the transposed matrix if it is not needed) and "outColumn" = {data0count}, the number of columns is extracted.

Alternatively, the second solution is to use the [loop](https://groops-devs.github.io/groops/html/loopType.html)->matrix again. I figured that the "inputfile", "transpose", "startRow" and "countRows" config elements define a sub-matrix (of the matrix defined by "inputfile") to which the [variable for data](https://groops-devs.github.io/groops/html/general.parser.html#dataVariables) (used to define the "variableLoop" config element) refer.

As a result, I managed to get the number of columns of the examined matrix by setting up [loop](https://groops-devs.github.io/groops/html/loopType.html)->matrix as follows:

- "inputfile" = (the file containing the matrix from which the columns will be extracted)
- "transpose" = "yes"
- "startRow" = 0
- "countRows" = "rows"
- the "variableLoop" config element "MatrixColumns" = "data0count" (or "rows")
- by using inside the loop the "FunctionsCalculate" program with the "statistics" config element set to "< enabled >", "outputfile" = (the file where the columns will be stored) and "outColumn" = {MatrixColumns}

A condition can also be set so as for the "FunctionsCalculate" program to be executed only at the end of the loop and thus avoid repeating it in every loop.

Then, the "outputfile" (either from the first or the second solution) and the file which contains as values the desired statistics from the examined matrix can be combined into one file by using again the "FunctionsCalculate" program.

Best regards,
Greg
